src/mesh/mesh_pack.hpp

```diff
@@ -0,0 +1,129 @@
+// (C) (or copyright) 2020. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//=======================================================================================
+#ifndef MESH_MESH_PACK_HPP_
+#define MESH_MESH_PACK_HPP_
+
+#include <array>
+#include <utility>
+
+#include "coordinates/coordinates.hpp"
+#include "interface/container.hpp"
+#include "interface/variable_pack.hpp"
+#include "kokkos_abstraction.hpp"
+#include "mesh/domain.hpp"
+#include "mesh/mesh.hpp" // TODO(JMM): Replace with forward declaration?
+
+namespace parthenon {
+
+// a separate dims array removes a branch case in `GetDim`
+// TODO(JMM): Using one IndexShape because its the same for all
+// meshblocks. This needs careful thought before sticking with it.
+template <typename T>
+class MeshPack {
+ public:
+  MeshPack() = default;
+  MeshPack(const ParArray1D<T> view, const IndexShape shape,
+           const ParArray1D<Coordinates_t> coordinates, const std::array<int, 5> dims)
+      : v_(view), cellbounds(shape), coords(coordinates), dims_(dims) {}
+  KOKKOS_FORCEINLINE_FUNCTION
+  auto &operator()(const int block) const { return v_(block); }
+  KOKKOS_FORCEINLINE_FUNCTION
+  auto &operator()(const int block, const int n) const { return v_(block)(n); }
+  KOKKOS_FORCEINLINE_FUNCTION
+  auto &operator()(const int block, const int n, const int k, const int j,
+                   const int i) const {
+    return v_(block)(n)(k, j, i);
+  }
+  KOKKOS_FORCEINLINE_FUNCTION
+  int GetDim(const int i) const {
+    assert(i > 0 && i < 6);
+    return dims_[i - 1];
+  }
+  KOKKOS_FORCEINLINE_FUNCTION
+  int GetNdim() const { return v_(0).GetNdim(); }
+  KOKKOS_FORCEINLINE_FUNCTION
+  int GetSparse(const int n) const { return v_(0).GetSparse(n); }
+
+  // TODO(JMM): Also include mesh domain object?
+  IndexShape cellbounds;
+  ParArray1D<Coordinates_t> coords;
+
+ private:
+  ParArray1D<T> v_;
+  std::array<int, 5> dims_;
+};
+
+template <typename T>
+using ViewOfPacks = ParArray1D<VariablePack<T>>;
+template <typename T>
+using ViewOfFluxPacks = ParArray1D<VariableFluxPack<T>>;
+template <typename T>
+using MeshVariablePack = MeshPack<VariablePack<T>>;
+template <typename T>
+using MeshVariableFluxPack = MeshPack<VariableFluxPack<T>>;
+
+// TODO(JMM): Should this be cached?
```

Comment:
I think the issue with caching with the mesh pack in its current form is that if you do the pack over the whole mesh, you would basically invalidate the cache every cycle. However, you could, I suppose, over allocate the pack sizes and grow the caches automatically. We'll probably? need caching, but it's an open question how  we do it.

Reply:
cache would only be invalidated when the meshblocks on a node change, so _possibly_ when AMR and/or load balancing are done.  In a block AMR code, that's typically not every cycle.

Reply:
Another factor: caching is already *kind* of implemented because `MeshPack` leverages the `VariablePack`, which is cached. I think we need to measure to see how much it hurts us.